### ISSUE TITLE
[RF] heap-use-after-free for RooAbsRealWrapper

### ISSUE THREAD

**Issue by hahnjo** [root-project/root#10009]
After #10005 is merged, a couple of tests will fail due to a `heap-use-after-free` or `stack-use-after-return`:
 * `gtest-roofit-roofitcore-test-testRooFitDriver` (`heap-use-after-free`)
 * `gtest-roofit-roofitcore-test-testRooAbsPdf` (`heap-use-after-free`)
 * `gtest-roofit-roofitcore-test-testTestStatistics` (`heap-use-after-free`)
 * `gtest-roofit-roofitcore-test-testLikelihoodSerial` (`heap-use-after-free`)
 * ~`gtest-roofit-roofit-test-testRooPoisson` (`stack-use-after-return`)~
 * `gtest-roofit-roofitcore-test-testSumW2Error` (`heap-use-after-free`)
 * `gtest-roofit-roostats-test-testSPlot` (`heap-use-after-free`)
 * ~`test-stressroofit` (`heap-use-after-free`)~
 * ~`test-stressroofit-interpreted` (`heap-use-after-free`)~
 * `tutorial-roofit-rf611_weightedfits` (`heap-use-after-free`)
 * ~`tutorial-roofit-rf705_linearmorph` (`stack-use-after-return`)~
 * ~`tutorial-roofit-rf705_linearmorph-py` (`stack-use-after-return`)~
 * `gtest-roottest-root-roofitstats-vectorisedPDFs-*` (`heap-use-after-free`)

The top of the stack trace (for the first failing test) is:
```
229: WRITE of size 8 at 0x61d000a37370 thread T0
229:     #0 0x7f5aa5163db3 in RooAbsArg::setDataToken(unsigned long) /home/jhahnfel/ROOT/src/roofit/roofitcore/inc/RooAbsArg.h:720:53
229:     #1 0x7f5aa5163db3 in ROOT::Experimental::NodeInfo::~NodeInfo() /home/jhahnfel/ROOT/src/roofit/roofitcore/src/RooFitDriver.cxx:100:15
229:     #2 0x7f5aa515d8f7 in void std::_Destroy<ROOT::Experimental::NodeInfo>(ROOT::Experimental::NodeInfo*) /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_construct.h:98:19                                                                                                                                                                                 
229:     #3 0x7f5aa515d8f7 in void std::_Destroy_aux<false>::__destroy<ROOT::Experimental::NodeInfo*>(ROOT::Experimental::NodeInfo*, ROOT::Experimental::NodeInfo*) /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_construct.h:108:6                                                                                                                           
229:     #4 0x7f5aa515d8f7 in void std::_Destroy<ROOT::Experimental::NodeInfo*>(ROOT::Experimental::NodeInfo*, ROOT::Experimental::NodeInfo*) /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_construct.h:136:7                                                                                                                                                 
229:     #5 0x7f5aa515d8f7 in void std::_Destroy<ROOT::Experimental::NodeInfo*, ROOT::Experimental::NodeInfo>(ROOT::Experimental::NodeInfo*, ROOT::Experimental::NodeInfo*, std::allocator<ROOT::Experimental::NodeInfo>&) /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_construct.h:206:7                                                                    
229:     #6 0x7f5aa515d8f7 in std::vector<ROOT::Experimental::NodeInfo, std::allocator<ROOT::Experimental::NodeInfo> >::~vector() /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_vector.h:567:2                                                                                                                                                                
229:     #7 0x7f5aa515d8f7 in ROOT::Experimental::RooFitDriver::~RooFitDriver() /home/jhahnfel/ROOT/src/roofit/roofitcore/src/RooFitDriver.cxx:252:1
229:     #8 0x7f5aa4cca57d in (anonymous namespace)::RooAbsRealWrapper::~RooAbsRealWrapper() /home/jhahnfel/ROOT/src/roofit/roofitcore/src/BatchModeHelpers.cxx:215:10
229:     #9 0x7f5aa4cca57d in (anonymous namespace)::RooAbsRealWrapper::~RooAbsRealWrapper() /home/jhahnfel/ROOT/src/roofit/roofitcore/src/BatchModeHelpers.cxx:213:4
229:     #10 0x37f28f in std::default_delete<RooAbsReal>::operator()(RooAbsReal*) const /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/unique_ptr.h:81:2
229:     #11 0x37f28f in std::unique_ptr<RooAbsReal, std::default_delete<RooAbsReal> >::~unique_ptr() /usr/bin/../lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/unique_ptr.h:277:4                                                                                                                                                                                            
229:     #12 0x37f28f in testRooFitDriver_SimpleLikelihoodFit_Test::TestBody() /home/jhahnfel/ROOT/src/roofit/roofitcore/test/testRooFitDriver.cxx:100:1
```

Note: There may be other problems in the tests after this issue is fixed. Please check locally with an instrumented build that the tests pass afterwards!

**Comment by hahnjo** [comment]
*Something* appears to have fixed *some* of the problems: `gtest-roofit-roofit-test-testSumW2Error` and all `gtest-roottest-root-roofitstats-vectorisedPDFs-*` pass now; `test-stressroofit` and `test-stressroofit-interpreted` instead run into a `stack-use-after-return`. `tutorial-roofit-rf611_weightedfits` times out, which it didn't before, so it *may* also be fixed...

**Comment by hahnjo** [comment]
I've updated the summary, there are definitely still problems in `RooAbsRealWrapper`. It would be really good if somebody could take a look, fix and close this issue and then make sure that it doesn't regress! There also seems to be a newer, much bigger problem for which I've opened https://github.com/root-project/root/issues/11186

**Comment by hahnjo** [comment]
@guitargeek great work! :tada:

There are still two failing tests (`gtest-roofit-roostats-test-testSPlot` and `tutorial-roofit-rf110_normintegration-py`) with the same initial symptoms that this issue was about, a `heap-use-after-free` in `RooAbsCollection::empty()`. Do you want me to open a new issue for the two? (`tutorial-roofit-rf104_classfactory-py` seems to fail for other reasons, not sure...)
